### PR TITLE
feat: Add support for HTTP keep-alive between the proxy and upstream

### LINKED ISSUE #609: How Do I Enable Keepalive connections
Please, what's the recommended way to enable `keepalive` connections.
I'm supposed to add something like
`keepalive 100;`
in the upstream directive, and 
`proxy_http_version 1.1;`
`proxy_set_header Connection "";`
in that of `location` correct?
But since the confs are generated, I can't seem to figure out a way to add these without breaking anything.

### PR DESCRIPTION
**Pull request by rhansen**
Fixes #609

Note: This depends on PR #2127, and will be marked as draft until that PR is merged.

### CODE REVIEW

**Comment by rhansen** [comment]
I'm not super satisfied with the tests I added, but I couldn't think of a good way to thoroughly test the feature (e.g., grepping the generated config file seems too fragile).

**Comment by rhansen** [comment]
Friendly ping @buchdag

**Comment by rhansen** [comment]
Marking as draft until #2127 is merged.

**Comment by buchdag** [comment]
@rhansen I've got mixed feeling about the use of labels. While I too do prefer labels to environment variables (as they inherently avoid stuff like #2139), I think we should stick to environment variables unless we're prepared to migrate everything to labels, as we don't use labels for anything yet. I also believe that people are more familiar with environment variables than they are with labels.

**Comment by rhansen** [comment]
> I think we should stick to environment variables unless we're prepared to migrate everything to labels, as we don't use labels for anything yet. I also believe that people are more familiar with environment variables than they are with labels.

Good points.  However, assuming we will eventually migrate the existing environment variables to labels (I think we should; they are designed for this purpose), using an environment variable for this feature makes more work for us now and later.  I'd rather live with the short-term inconsistency.

**Comment by buchdag** [comment]
@rhansen the way I see it, in the end we should keep environment variables for configuration of the nginx-proxy (or docker-gen) container, and switch to labels on the proxied containers. Do you agree with this ?

**Comment by rhansen** [comment]
I'm not sure.  There are pros and cons either way.  Advantages of environment variables for the `nginx-proxy` container:
  * Less work: we don't need to migrate them to labels like we do with the environment variables on proxied containers.
  * Having distinct ways of configuring `nginx-proxy` vs. the other containers might make configs more readable (environment variables are for the container's own use, labels are for use by a different container).

Advantages of labels:
  * Consistency: all configuration is done via labels (and include files).
  * Less risk of collision with environment variables used by other code in the `nginx-proxy` container (nginx itself, a custom nginx module added by the user, the shell, `docker-gen`, Docker API libraries, logging libraries, etc.).
  * Less pollution of the environment seen by various processes.
  * It might be easier to prevent undesired disclosure of privacy- or security-sensitive values:  Environment variables are normally copied to every spawned process, which makes it easier for an attacker to find a way to inspect the configuration.  It should be easier to lock down access to the Docker socket to prevent such disclosure.  (I haven't carefully studied the security implications so labels might actually increase the risk.)

**Comment by rhansen** [comment]
I opened #2148 for further discussion about environment variables vs. labels.

**Comment by buchdag** [comment]
@rhansen I'll review this again tonight or  tomorrow but it should be okay to merge before next release.

**Comment by buchdag** [comment]
@rhansen I did not notice the experimental status of the feature. What would be needed to confirm that it work as intended ?

**Comment by rhansen** [comment]
> What would be needed to confirm that it work as intended ?

Time, or discussions/bug reports.  We don't have a way of collecting usage statistics, so there's not much we can do other than eventually think to ourselves, "this feature has been around for a while and nobody has complained, so it's probably OK."

**Comment by SchoNie** [comment]
keepalive is not experimental and is even [recommended to have enabled.](https://www.nginx.com/blog/avoiding-top-10-nginx-configuration-mistakes/#no-keepalives) I do like the way of this implementation by @rhansen (also I like the docker label concept)
Currently I manually overwrite the template file to have the keepalive directions in there. So having this merged would help me to run this container clean without any overrides.

**Comment by rhansen** [comment]
> keepalive is not experimental

Agreed; the experimental thing here is the *ability to enable* keepalive in `nginx-proxy`, not keepalive itself. 

> I do like the way of this implementation by @rhansen (also I like the docker label concept)

Thanks for the feedback!

**Comment by rhansen** [comment]
> I suspect there is a problem setting the `proxy_set_header Connection $proxy_connection_noclose;` in the location block.

Good catch, thank you @SchoNie!

> Maybe a better way would be to wrap the map in an if, and overwrite the first $proxy_connection (it will stay upgrade if received, else it clears the close with an empty string).

Thanks for the suggestion.  I went with a different approach that preserves per-upstream configurability (it can be turned on for some containers and left off for others).  I also updated the tests to check for the regression you found.  If you have the time, I would appreciate it if you looked at the new revision.

**Comment by SchoNie** [comment]
Tested, LGTM.
Good you added a X-Real-IP test!

File: README.md
Comment (buchdag) [inline README.md:379]: Could you open a discussion to provide feedback on this feature and put a link to this discussion here ? I think it can be merged once this is done.

**Review by buchdag — CHANGES_REQUESTED** [review]

File: README.md
Comment (rhansen) [reply README.md]: Done.

**Review by buchdag — APPROVED** [review]

**Comment by buchdag** [comment]
@rhansen could you just wait a few days before merging this ? I'd like to release a patch version of docker-gen and then of nginx-proxy before merging this feature and the load balancing one from @SchoNie

**Pull request merged**
